### PR TITLE
Enhance database config templates for flexible env connections

### PR DESCRIPTION
**Pull request by viktorianer**
### Motivation / Background

This Pull Request has been created to enhance the flexibility and usability of the Rails `database.yml` configuration for PostgreSQL databases. Currently, Rails applications default to connecting to PostgreSQL through a local Unix socket for development environments. This setup works well for local development but becomes cumbersome when developers use Dockerized PostgreSQL instances or switch between local and Docker environments. 

```bash
> rails db:prepare
bin/rails aborted!
ActiveRecord::ConnectionNotEstablished: connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory (ActiveRecord::ConnectionNotEstablished)
	Is the server running locally and accepting connections on that socket?

Caused by:
PG::ConnectionBad: connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory (PG::ConnectionBad)
	Is the server running locally and accepting connections on that socket?
```

By facilitating easy switching between these environments without the need to manually edit the `database.yml` file, this PR aims to streamline the development workflow, making it more efficient and less error-prone.

### Detail

This Pull Request changes the PostgreSQL database configuration template (`postgresql.yml.tt`) by introducing dynamic environment variable support for the `host`, `username`, and, `password` settings in the default configuration. Specifically, it allows these settings to be overridden via environment variables, with defaults provided for local development. This approach enables developers to easily switch between a local PostgreSQL server and a Dockerized PostgreSQL instance by simply setting or unsetting environment variables, without the need for manual configuration file edits.

### Additional information

This approach supports the idea in https://github.com/rails/rails/pull/50914#discussion_r1478343297.

~~I do not like to use `"<%= app_name.upcase %>_..." in variable names, but I followed the naming in https://github.com/rails/rails/blob/2c5fe6fa7aba4a10d56ff4ee54d26511e58dda3b/railties/lib/rails/generators/rails/app/templates/config/databases/postgresql.yml.tt#L84.~~

~~As suggested in https://github.com/rails/rails/pull/50914#discussion_r1478855943, I would rather use `DATABASE_URL`, `POSTGRES_USER`, and `POSTGRES_PASSWORD`.~~

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

I am not able to run the tests in Dev Container or locally on macOS (it crashes 👎🏼).  Of course, I would love to add or fix tests, need more tome to set up the local environment.

But I could create a new app from local branch, which includes correct generated Postgres database config (https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#running-an-application-against-your-local-branch).

### CODE REVIEW

**Comment by palkan** [comment]
> I would rather use DATABASE_URL

So, why don't we go this way? What I usually do is adding a single line to the `database.yml`:

```yml
default: &default
  adapter: postgresql
  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
  url: <%= ENV["DATABASE_URL"] %> # <--- THIS
  timeout: 5000
```

If the URL is defined in the YAML configuration, it wins (I'm not sure but we simply put it in the env and have host/user configuration in the YAML, the env one wouldn't be used as the primary source of configuration).

I don't see any benefit of having multiple env vars.

**Comment by viktorianer** [comment]
> but we simply put it in the env.

Thank you for your suggestion regarding the use of `DATABASE_URL`. While this approach simplifies the configuration by consolidating it into a single environment variable, it may not fully address the flexibility required for workflows across multiple development environments.

The essence of the proposed change in the PR is to enhance flexibility and allow developers to seamlessly switch between different PostgreSQL setups—be it `local`, `Dockerized`, or within a `Dev Container`— **without the need to alter their project configuration**. This is particularly important in the early stages of a project or when **developers have varying preferences for their development setup**.

Consider the following scenario, which is quite common among Rails developers:

1. Starting a New Project with Dockerized PostgreSQL

```bash
$ rails new <project_name> --database=postgresql
$ docker run -it --rm -d -p 5432:5432 -e POSTGRES_PASSWORD=postgres --name postgres postgres:latest
$ bin/setup
...
bin/rails aborted!
== Preparing database ==
bin/rails aborted!
ActiveRecord::ConnectionNotEstablished: connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory (ActiveRecord::ConnectionNotEstablished)
	Is the server running locally and accepting connections on that socket?
...
```

In this case, the developer expects to connect to PostgreSQL running in a Docker container without further configuration changes. However, they encounter an error because Rails defaults to expecting a local PostgreSQL server, connecting through a Unix socket (the last time I used Unix socket in my _development environment_ is around 6 years ago!).

2. Switching Environments

How does one then switch from a Dockerized PostgreSQL to a database running inside a Dev Container, or to a local-based PostgreSQL instance? 
The `DATABASE_URL` approach requires manual updates to this variable across different environments.

Furthermore, the behavior of Rails database tasks can differ when the `url` option is present in the `database.yml` file. This can introduce inconsistencies in how migrations, setups, and resets are handled, **complicating the development process rather than simplifying it**.

By allowing the configuration to dynamically adapt based on environment variables for `host`, `username`, `password`, and `port`, we cater to a broader range of development practices and preferences.

### In summary

- The `DATABASE_URL` configuration works well for workflows with inside a Dev Container (Docker Compose), or to a cloud-based PostgreSQL (staging, production). 
- The proposed changes aim to provide a more versatile and developer-friendly solution to database configuration management across multiple environments without necessitating manual adjustments or relying on external tooling for environment management.

**Comment by matthewd** [comment]
> provide a more versatile and developer-friendly solution to database configuration management across multiple environments without necessitating manual adjustments

It's a configuration file. The point of it existing is to facilitate manual adjustments. Why would we want to ship a config file [which configures our own code] that merely says "actually look over here" for every key?

**Comment by viktorianer** [comment]
> > provide a more versatile and developer-friendly solution to database configuration management across multiple environments without necessitating manual adjustments
> 
> It's a configuration file. The point of it existing is to facilitate manual adjustments. Why would we want to ship a config file [which configures our own code] that merely says "actually look over here" for every key?

Thank you for your feedback and for engaging with the proposal. 
The essence of this proposal, however, is not to detract from the configurability offered by `database.yml` (https://github.com/rails/rails/blob/v7.1.3/railties/lib/rails/generators/rails/app/templates/config/databases/postgresql.yml.tt#L84), but to enhance it in a way that aligns with modern development practices (Docker, remote containers, etc.) and workflows (switching between local and remote).

### Simplified Change

What if we change the line to simply hardcode the username as "postgres":

```yml
username: postgres
...
password: postgres
...
host: localhost
```

The primary goal is to _simplify the initial setup for new Rails applications_ while accommodating the common development practice of using the "postgres" user for local and Docker-based PostgreSQL instances.

> By allowing the configuration to dynamically adapt based on environment variables for `host`, `username`, `password`, and `port`, we cater to a broader range of development practices and preferences.

In order to archive simplified initial setup, I think at least, we need to move the 4 setting options from `development:` section to `default:`. _In this case, we can leave it even uncommented?!_

However, it's essential to weigh this against the loss of flexibility and the potential need for manual configuration adjustments in environments where the "postgres" username is not used. Of course, In those cases, we can still use `DATABASE_URL`, see https://github.com/rails/rails/blob/v7.1.3/railties/lib/rails/generators/rails/app/templates/config/databases/postgresql.yml.tt#L75.

**Comment by rafaelfranca** [comment]
Thank you for the pull request but I don't think all this complexity is necessary by default. You are planning to setup a docker environment you can just edit your files in your application. If you are using devcontainers, Rails will do this for you in #50914.

We have no plan to support every single possible container development in Rails. Only devcontainers.
